annotation_pipeline/molecular_db.py

```diff
@@ -72,13 +72,13 @@ def _store(segm_i):
 
     pw = pywren.ibm_cf_executor(config=config, runtime_memory=2048)
     futures = pw.map(generate_formulas, adducts)
-    segments_n = set().union(*pw.get_result(futures))
+    segments_n = list(set().union(*pw.get_result(futures)))
     append_pywren_stats(futures, pw.config['pywren']['runtime_memory'])
 
     def deduplicate_formulas_segment(segm_i, ibm_cos, clean=True):
         print(f'Deduplicating formulas segment {segm_i}')
         objs = ibm_cos.list_objects_v2(Bucket=bucket, Prefix=f'{formulas_chunks_prefix}/chunk/{segm_i}/')
-        keys = [obj['Key'] for obj in objs['Contents']]
+        keys = [obj['Key'] for obj in objs['Contents']] if 'Contents' in objs else []
```

Comment:
`obj` should always contain `'Contents'`

Reply:
It doesn't contain `'Contents'` when there are no objects matching the prefix.

```
> de_client.list_objects_v2(Bucket='metaspace-lachlan-public', Prefix='something-that-doesnt-exist')

{'ResponseMetadata': {'RequestId': '58ac23f5-ca98-45eb-8e1f-a12a2a5c6fad',
  'HostId': '',
  'HTTPStatusCode': 200,
  'HTTPHeaders': {'date': 'Tue, 10 Dec 2019 10:44:49 GMT',
   'x-clv-request-id': '58ac23f5-ca98-45eb-8e1f-a12a2a5c6fad',
   'server': 'Cleversafe/3.14.7.88',
   'x-clv-s3-version': '2.5',
   'accept-ranges': 'bytes',
   'x-amz-request-id': '58ac23f5-ca98-45eb-8e1f-a12a2a5c6fad',
   'ibm-sse-kp-enabled': 'false',
   'content-type': 'application/xml',
   'content-length': '321'},
  'RetryAttempts': 0},
 'IsTruncated': False,
 'Name': 'metaspace-lachlan-public',
 'Prefix': 'something-that-doesnt-exist',
 'Delimiter': '',
 'MaxKeys': 1000,
 'KeyCount': 0}
```

I'm not sure why `deduplicate_formulas_segment` was looking for a missing prefix. I had removed a bunch of modifiers and databases from the `input_config.json` that I was using, but it still included enough formulas for `check_results` to report that the results were correct after this change.

Reply:
I meant that, at this stage, all keys should exist and that's why `obj` should always contain `'Contents'`, so I think that it isn't necessary to check if `obj` contains `'Contents'`.
I haven't understood the next part, did you find something wrong in the behavior of `deduplicate_formulas_segment` ?